### PR TITLE
Centroid calculation

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

File: src/algorithm/centroid.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        let p = Point(c);
+        let vec = Vec::new();
+        let linestring = LineString(vec);
+        let centroid = linestring.centroid();
+
+        assert_eq!(p.lat(), centroid.lat());
+        assert_eq!(p.lng(), centroid.lng());
+    }
+    #[test]
+    fn linestring_one_point_test() {
```
Comment (kinnou02) [inline src/algorithm/centroid.rs]: A test with an empty line string would be nice.

File: examples/types.rs [hunk 2]
```diff
@@ -12,4 +13,9 @@ fn main() {
 
     let Point(coord) = p;
     println!("Point at ({}, {})", coord.x, coord.y);
+    
+    let mut vec = Vec::new();
+    vec.push(p);
+    let linestring = LineString(vec);
```
Comment (kinnou02) [inline examples/types.rs]: Maybe you can create a new example file

File: src/algorithm/centroid.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+pub trait Centroid {
+    fn centroid(&self) -> Point;
+}
+
+impl Centroid for LineString {
+    fn centroid(&self) -> Point {
+        let sum = self.0
+                      .iter()
+                      .fold(Point(Coordinate { x: 0., y: 0. }), |sum, &p| sum + p);
+        let size = self.0.len() as f64;
```
Comment (antoine-de) [inline src/algorithm/centroid.rs]: hum I think the casting should be done only in the else (maybe directly in the  `/`: `x: sum.lng() / size as f64,` )

**Comment by kinnou02** [comment]
A little rebase and it's good to go!

**Comment by TeXitoi** [comment]
But, centroid on a LineString is not the mean of the points. Imagine:

```
+---------------------------------------+-+-+-+-+-+
                  centroid ^           ^ mean
```

**Comment by TeXitoi** [comment]
I think that's the mean of the middle of the segment weighted by the length of the segments.

File: src/algorithm/centroid.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
+pub use types::{Coordinate, Point, LineString};
+
+pub trait Centroid {
+    fn centroid(&self) -> Result<Point, &'static str>;
+}
+
+impl Centroid for LineString {
+    fn centroid(&self) -> Result<Point, &'static str> {
+        let vect = &self.0;
+        if vect.len() == 0 {
```
Comment (antoine-de) [inline src/algorithm/centroid.rs]: isn't there an `is_empty()` method ?

Comment (azime) [inline src/algorithm/centroid.rs]: OK

**Comment by azime** [comment]
@TeXitoi, i modified on your comments

File: src/algorithm/centroid.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
+pub use types::{Coordinate, Point, LineString};
+
+pub trait Centroid {
+	/// Calculation the centroid, see: https://en.wikipedia.org/wiki/Centroid
+    fn centroid(&self) -> Result<Point, &'static str>;
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: As the error of the centroid calculation seems trivial, what about just returning an Option?

File: src/algorithm/centroid.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    /// +---------------------------------------+-+-+-+-+-+
+    ///                  centroid ^           ^ mean
+    ///
+    fn centroid(&self) -> Result<Point, &'static str> {
+        let vect = &self.0;
+        if vect.is_empty() {
+            return Err("Empty linestring");
+        }
+        if vect.len() == 1 {
+            return Ok(Point(Coordinate {
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: useless return, or just remove the else to limit indentation, as you prefere.

File: src/algorithm/centroid.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    fn linestring_test() {
+        // linestring
+        let mut vect = Vec::new();
+        vect.push(Point(Coordinate { x: 0., y: 3. }));
+        vect.push(Point(Coordinate { x: 1., y: 1. }));
+        vect.push(Point(Coordinate { x: 2., y: 0. }));
+        vect.push(Point(Coordinate { x: 6., y: 0. }));
+        vect.push(Point(Coordinate { x: 7., y: 8. }));
+        vect.push(Point(Coordinate { x: 9., y: 8. }));
+        vect.push(Point(Coordinate { x: 10., y: 6. }));
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]:
``` rust
let p = |x, y| Point(Coordinate { x: x, y: y });
let vect = vec![p(0., 3.), p(1., 1.), ...];
```

File: src/algorithm/centroid.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        let linestring = LineString(vect);
+        let centroid = linestring.centroid();
+        assert_eq!(centroid.unwrap().lat(), 3.4629359659988497);
+        assert_eq!(centroid.unwrap().lng(), 5.458359775816375);
+    }
+}
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs:132]: Could be great to have a test as I proposed:

``` rust
let p = |x| Point(Coordinate  { x: x, y: 1. });
let linestring = LineString(vec![p(1.), p(7.), p(8.), p(9.), p(10.), p(11.)]);
assert_eq!(linestring.centroid(), Some(Point(Coordinate  { x: 6., y: 1. }));
```

File: src/algorithm/centroid.rs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            }));
+        } else {
+            let mut sum_x = 0.;
+            let mut sum_y = 0.;
+            let mut total_length = 0.;
+            for i in 0..vect.len() - 1 {
+                let segment_len = vect[i].distance_to(&vect[i + 1]);
+                total_length = total_length + segment_len;
+                sum_x = sum_x + segment_len * ((vect[i].lng() + vect[i + 1].lng()) / 2.);
+                sum_y = sum_y + segment_len * ((vect[i].lat() + vect[i + 1].lat()) / 2.);
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: you can use `+=`

File: src/algorithm/centroid.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        if vect.len() == 1 {
+            return Ok(Point(Coordinate {
+                x: vect[0].lng(),
+                y: vect[0].lat(),
+            }));
+        } else {
+            let mut sum_x = 0.;
+            let mut sum_y = 0.;
+            let mut total_length = 0.;
+            for i in 0..vect.len() - 1 {
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]:
``` rust
for (a, b) in v.iter().zip(v[1..].iter()) {
}
```

Comment (TeXitoi) [inline src/algorithm/centroid.rs]: Better to avoid `v[i]` when you can (it can panic).

File: src/algorithm/centroid.rs [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+pub trait Centroid {
+    /// Calculation the centroid, see: https://en.wikipedia.org/wiki/Centroid
+    fn centroid(&self) -> Result<Point, &'static str>;
+}
+
+impl Centroid for LineString {
+    /// Centroid on a LineString is the mean of the middle of the segment
+    /// weighted by the length of the segments:
+    /// +---------------------------------------+-+-+-+-+-+
+    ///                  centroid ^           ^ mean
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: I think the picture is not so clear. Maybe just remove it.

File: src/algorithm/centroid.rs (hunk 10 above)
Comment (antoine-de) [inline src/algorithm/centroid.rs]: I thought I remember an itertools stuff to do this, but I only found this:

https://github.com/slapresta/rust-iterslide

but It returns a list and not tuples :disappointed:

File: src/algorithm/centroid.rs (hunk 9 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

File: src/algorithm/centroid.rs (hunk 11 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

File: src/algorithm/centroid.rs (hunk 8 above)
Comment (azime) [inline src/algorithm/centroid.rs:132]: ok

File: src/algorithm/centroid.rs (hunk 10 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

File: src/types.rs [hunk 12]
```diff
@@ -153,6 +153,19 @@ impl Point {
...
     }
+    /// Returns the distance between two points:
+    ///
+    /// ```
+    /// use geo::Point;
+    ///
+    /// let p = Point::new(-72.1235, 42.3521);
+    /// let dist = p.distance_to(&Point::new(-72.1260, 42.45));
+    ///
+    /// assert_eq!(dist, 0.09793191512474639);
```
Comment (antoine-de) [inline src/types.rs]: maybe you shouldn't directly compare float there  but round them a bit

File: src/algorithm/centroid.rs (hunk 6 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

File: src/algorithm/centroid.rs (hunk 7 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

File: src/types.rs (hunk 12 above)
Comment (azime) [inline src/types.rs]: OK

**Comment by azime** [comment]
@TeXitoi, i modified on your comments

File: src/algorithm/centroid.rs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    /// Centroid on a LineString is the mean of the middle of the segment
+    /// weighted by the length of the segments.
+    ///
+    fn centroid(&self) -> Result<Point, &'static str> {
+        let vect = &self.0;
+        if vect.is_empty() {
+            return Err("Empty linestring");
+        }
+        if vect.len() == 1 {
+            Ok(Point(Coordinate {
```
Comment (antoine-de) [inline src/algorithm/centroid.rs]: I think it would be better to clone the coordinate

File: src/algorithm/centroid.rs [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+mod test {
+    use types::{Coordinate, Point, LineString, Polygon};
+    use algorithm::centroid::Centroid;
+    #[test]
+    fn empty_linestring_test() {
+        // Empty linestring
+        let vec = Vec::new();
+        let linestring = LineString(vec);
+        let centroid = linestring.centroid();
+        assert_eq!(centroid.is_none(), true);
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: just assert!()

File: src/algorithm/centroid.rs [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        let p = Point::new(40.02f64, 116.34);
+        let mut vect = Vec::new();
+        vect.push(p);
+        let linestring = LineString(vect);
+        let centroid = linestring.centroid();
+        assert_eq!(centroid, Some(p));
+    }
+    #[test]
+    fn linestring_test() {
+        // linestring
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs]: useless comment you can do a ascii art if you want to explain the test

File: src/algorithm/centroid.rs [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        let p = |x, y| Point(Coordinate { x: x, y: y });
+        let v = Vec::new();
+        let linestring = LineString(vec![p(0., 0.), p(1., 0.), p(2., 1.), p(3., 1.), p(3., 2.),
+                                         p(2., 3.), p(1., 3.), p(1., 4.), p(0., 4.), p(0., 0.)]);
+        let poly = Polygon(linestring, v);
+        let distance = poly.centroid().unwrap().distance_to(&p(1.1666666666666667, 1.875));
+        // distance = 0
+        assert!(distance < 1e-6);
+    }
```
Comment (TeXitoi) [inline src/algorithm/centroid.rs:131]: maybe we can add a user understandable test, as the centroid of a square

File: src/types.rs [hunk 17]
```diff
@@ -153,6 +153,19 @@ impl Point {
...
+    /// use geo::Point;
+    ///
+    /// let p = Point::new(-72.1235, 42.3521);
+    /// let dist = p.distance_to(&Point::new(-72.1260, 42.45));
+    ///
+    /// assert!(dist < 1e-1)
+    /// ```
+    pub fn distance_to(&self, point: &Point) -> f64 {
+        ((self.x() - point.x()).powi(2) + (self.y() - point.y()).powi(2)).sqrt()
+    }
```
Comment (TeXitoi) [inline src/types.rs:168]: no unit tests?

Comment (azime) [inline src/types.rs:168]: OK

File: src/algorithm/centroid.rs (hunk 16 above)
Comment (azime) [inline src/algorithm/centroid.rs:131]: OK

File: src/algorithm/centroid.rs (hunk 15 above)
Comment (azime) [inline src/algorithm/centroid.rs]: ok

File: src/algorithm/centroid.rs (hunk 14 above)
Comment (azime) [inline src/algorithm/centroid.rs]: OK

**Comment by azime** [comment]
@TeXitoi, i modified on your comments

File: src/algorithm/centroid.rs [hunk 18]
```diff
@@ -1,7 +1,23 @@
+
 pub use types::{Coordinate, Point, LineString, Polygon};
 
+/// Calculation the centroid.
```
Comment (kinnou02) [inline src/algorithm/centroid.rs]: of

File: src/algorithm/centroid.rs [hunk 19]
```diff
@@ -38,7 +54,7 @@ impl Centroid for Polygon {
     /// See: https://en.wikipedia.org/wiki/Centroid
     ///
     fn centroid(&self) -> Option<Point> {
-        // TODO: consideration of Vec<LineString>);
+        // TODO: consideration of inner polygon;
```
Comment (kinnou02) [inline src/algorithm/centroid.rs]: polygons

**Pull request merged**